text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+
+2. When type checking an expression `match e_s { a_1 => e_1, ... a_n => e_n }`,
+   for each match arm `a_i` which contains a pattern of form `p_i | q_i`,
+   the pattern `p_i | q_i` is considered ill formed if,
+   at the depth `d` where it exists the fragment of `e_s` at depth `d`,
+   the type of the expression fragment does not unify with `p_i | q_i`.
+
+3. With respect to exhaustiveness checking, a pattern `p | q` is
+   considered to cover `p` as well as `q`. For some constructor `c(x, ..)`
+   the distributive law applies such that `c(p | q, ..rest)` covers the same
+   set of value as `c(p, ..rest) | c(q, ..rest)` does. This can be applied
+   recursively until there are no more nested patterns of form `p | q` other
+   than those that exist at the top level.
+
+   Note that by *"constructor"* we do not refer to tuple struct patterns,
+   but rather we refer to a pattern for any product type.
+   This includes enum variants, tuple structs, structs with named fields,
+   arrays, tuples, and slices.
+
+## Dynamic semantics
+
+1. The dynamic semantics of pattern matching a scrutinee expression `e_s`
+   against a pattern `c(p | q, ..rest)` at depth `d` where `c` is some constructor,
+   `p` and `q` are arbitrary patterns, and `rest` is optionally any remaining
+   potential factors in `c`, is defined as being the same as that of
+   `c(p, ..rest) | c(q, ..rest)`.
+
+## Implementation notes
+
+With respect to both static and dynamic semantics,
+it is always valid to first desugar a pattern `c(p | q)`
+in CNF to its equivalent form in DNF, i.e. `c(p) | c(q)`.
+
+However, implementing `c(p | q)` in terms of a pure desugaring to `c(p) | c(q)`
+may not be optimal as the desugaring can result in exponential blow-up of patterns.
+Instead, it is more likely that a one-step case analysis will be more efficient.
+
+Which implementation technique to use is left open to each Rust compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Some parsers will have to be rewritten by a tiny bit;
+   We do this with any syntactic change in the language so
+   there should not be any problem.
+
+# Rationale and alternatives
+[alternatives]: #rationale-and-alternatives
+
+As for why the change as proposed in this RFC should be done,
+it is discussed in the [motivation].
+
+Since we already use `|` for alternation at the top level, the only consistent
+operator syntax for alternations in nested patterns would be `|`.
+Therefore, there are not many design choices to make with respect to *how*
+this change should be done rather than *if*.
+
+With respect to the precedence of `|`, we already cannot interpret `i @ p | q`
+as `i @ (p | q)` because it is already legal to write `i @ p | j @ q`.
+Therefore, if we say that `|` binds more tightly,
+then `i @ p | j @ q` will associate as `i @ (p | j @ q)` which as a different
+meaning than what we currently have, thus causing a breaking change.
```

Comment:
OK, fair enough, this makes sense.

However, could you please specify a lint and suggestion for this? Specifically, if you write `i @ p | q`, you will already get an error because the two patterns (`i@p` and `q`) don't both bind `i`, so with a bit of extra error-reporting logic we could add a suggestion for `i @ (p | q)`.